### PR TITLE
Add Tooltip component and radix-ui react-tooltip

### PR DESCRIPTION
**Pull request by Oksamies**
As the title says

refs TS-1315

### CODE REVIEW

File: packages/cyberstorm/src/components/Tooltip.tsx [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  // const { label, tooltipStyle } = props;
+  return (
+    <RadixTooltip.Provider delayDuration={150}>
+      <RadixTooltip.Root>
+        <RadixTooltip.Trigger>{children}</RadixTooltip.Trigger>
+        <RadixTooltip.Portal>
+          <RadixTooltip.Content
+            /* TS is not aware of defaultProps of function components. */
+            /* eslint-disable-next-line @typescript-eslint/no-non-null-assertion */
+            className={`${styles.tooltip} ${styles[tooltipStyle!]}`}
```
Comment (MythicManiac) [inline packages/cyberstorm/src/components/Tooltip.tsx:26]: As discussed in Discord, this and other similar usages have to be changed. A side benefit is that you can get rid of the eslint disabling comments.

File: packages/cyberstorm/src/components/componentStyles/Tooltip.module.css [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+button {
+  all: unset;
+}
+
+.tooltip {
+  padding: 0.5rem 0.75rem;
+  border-radius: 0.5rem;
+  color: var(--color-white);
+
+  font-weight: 700;
```
Comment (VilppeRiskidev) [inline packages/cyberstorm/src/components/componentStyles/Tooltip.module.css:11]: I think the whole file needs refactoring when the global.css changes drop from the PackageCard PR. Use variables and better naming and stuff.

File: apps/cyberstorm-storybook/stories/components/Tooltip.stories.tsx [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+      <Tooltip {...args}>
+        <Button {...buttonargs} />
+      </Tooltip>
+    </div>
+  </div>
+);
+
+const DefaultTooltip = Template.bind({});
+DefaultTooltip.args = defaultArgs;
+
```
Comment (VilppeRiskidev) [inline apps/cyberstorm-storybook/stories/components/Tooltip.stories.tsx:62]: Could probably use another exmple of what happens when basicly no props are passed. as in empty parameters.
The pattern is:
MinimalTooltip (minimal to no props passed)
ReferenceTooltip (matching the design)
ExtremeTooltip ("stupid" extreme case, for example 500 chars long label etc.)

Comment (VilppeRiskidev) [reply apps/cyberstorm-storybook/stories/components/Tooltip.stories.tsx:30]: What I meant by Minimal was to (in this case) leave the props empty, so it acts as an example of what happens when no label is passed. Since the label is an optional prop.

Comment (VilppeRiskidev) [reply apps/cyberstorm-storybook/stories/components/Tooltip.stories.tsx:30]: Also the long one could be even longer, like 200 chars or something :D

File: packages/cyberstorm/src/components/componentStyles/Tooltip.module.css [hunk 4]
```diff
@@ -0,0 +1,84 @@
+.tooltip {
+  padding: 0.5rem 0.75rem;
+  border-radius: 0.5rem;
+  color: var(--color-white);
+
+  font-weight: var(--font-weight-bold);
+  font-size: var(--size-m);
+  background-color: var(--color-default-purple--4);
```
Comment (VilppeRiskidev) [inline packages/cyberstorm/src/components/componentStyles/Tooltip.module.css:8]: I think this one should use the surface variable `--color-surface--3`

**Review by VilppeRiskidev — CHANGES_REQUESTED** [review]
Mainly needs more complying to the naming conventions https://github.com/thunderstore-io/thunderstore-ui/blob/master/packages/cyberstorm/README.md

File: packages/cyberstorm/src/components/Tooltip.tsx [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+// Cyberstorm Tooltip component
+
+export const Tooltip: React.FC<PropsWithChildren<TooltipProps>> = ({
+  label,
+  children,
+}) => {
+  return (
+    <RadixTooltip.Provider delayDuration={150}>
```
Comment (MythicManiac) [inline packages/cyberstorm/src/components/Tooltip.tsx:16]: Is this provider something that has to be included for every tooltip individually, or should it be included once globally?

File: packages/cyberstorm/src/components/Tooltip.tsx [hunk 6]
```diff
@@ -0,0 +1,35 @@
+import React, { PropsWithChildren } from "react";
+import * as RadixTooltip from "@radix-ui/react-tooltip";
+import styles from "./componentStyles/Tooltip.module.css";
+
+export interface TooltipProps {
+  label?: string;
```
Comment (MythicManiac) [inline packages/cyberstorm/src/components/Tooltip.tsx:6]: Could this be `ReactNode`? I don't think we'll only restrict tooltips to simple strings.

File: packages/cyberstorm/src/components/Tooltip.tsx [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+        <RadixTooltip.Portal>
+          <RadixTooltip.Content
+            className={`${styles.tooltip}`}
+            sideOffset={5}
+            side={"right"}
+            collisionPadding={20}
+            sticky={"always"}
+            avoidCollisions={true}
+          >
+            {label}
```
[the remark this answers is not in the document]
Comment (MythicManiac) [reply packages/cyberstorm/src/components/Tooltip.tsx:28]: This is actually not necessary here since react doesn't render falsy values. `label ? label : null` is evaluated in the same way as just `label` but with a bit of extra steps.

**Review by MythicManiac — COMMENTED** [review]
Not much else to add on top of @VilppeRiskidev's comments

File: apps/cyberstorm-storybook/stories/components/Tooltip.stories.tsx (hunk 3 above)
Comment (Oksamies) [reply apps/cyberstorm-storybook/stories/components/Tooltip.stories.tsx:30]: Done and done. Although I'm not sure why the text doesn't get wrapped to another line with the Radix's Tooltip. It just goes off the screen. `overflow-wrap` doesn't help.

File: packages/cyberstorm/src/components/Tooltip.tsx (hunk 5 above)
Comment (Oksamies) [reply packages/cyberstorm/src/components/Tooltip.tsx:16]: It at least doesn't work without it. I think it can be defined once globally, but the question is; do want to do that? Cause then the app that's using CS would need to define the provider. If I've understood correctly. Let me know if there is a way to do this, that I'm not aware of.

File: packages/cyberstorm/src/components/Tooltip.tsx [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+}) => {
+  return (
+    <RadixTooltip.Provider delayDuration={150}>
+      <RadixTooltip.Root>
+        <RadixTooltip.Trigger>{children}</RadixTooltip.Trigger>
+        <RadixTooltip.Portal>
+          <RadixTooltip.Content
+            className={`${styles.tooltip}`}
+            sideOffset={5}
+            side={"right"}
```
[the remark this answers is not in the document]
Comment (Oksamies) [reply packages/cyberstorm/src/components/Tooltip.tsx:23]: I've added all of the Radix Tooltips props, as props to our Tooltip component. Can't see why I wouldn't do that. Good suggestion 👍

File: apps/cyberstorm-storybook/stories/components/Tooltip.stories.tsx (hunk 3 above)
Comment (gkurck) [reply apps/cyberstorm-storybook/stories/components/Tooltip.stories.tsx:30]: ^ This is because there aren't any constraints to enable wrapping of the content, because:

1) The tooltip is rendered through portal to body, so the parent isn't constraining the width
2) It's has fixed positioning
3) There's no max-width set for the tooltip

**Review by VilppeRiskidev — APPROVED** [review]
Looking good to me, as long as other reviewers' comments are addressed.

File: packages/cyberstorm/src/components/Tooltip.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (VilppeRiskidev) [reply packages/cyberstorm/src/components/Tooltip.tsx]: Very nice!

File: packages/cyberstorm/src/components/Tooltip.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (VilppeRiskidev) [reply packages/cyberstorm/src/components/Tooltip.tsx]: Oh, nice to know! so only need to do it when the value is wrapped in some element etc.👍

File: apps/cyberstorm-storybook/stories/components/Tooltip.stories.tsx (hunk 3 above)
Comment (VilppeRiskidev) [reply apps/cyberstorm-storybook/stories/components/Tooltip.stories.tsx:30]: Oksamies, feel free to resolve once you've read gkurck's comment

File: packages/cyberstorm/src/components/Tooltip.tsx (hunk 5 above)
Comment (MythicManiac) [reply packages/cyberstorm/src/components/Tooltip.tsx:26]: > I think it can be defined once globally, but the question is; do want to do that?

That is the question I'm asking. If it can be defined once globally without negative side effects to it's usage, then yes we should do that as it's more efficient. However, it could also be that the API of Radix requires that each tooltip has an unique context.

So refer to the docs and/or test around. If a single global one works, that should be used.

> Cause then the app that's using CS would need to define the provider.

Yeah, this is standard practice; we will have multiple providers at the root level of all apps in every case. What we might want to do is wrap them all into a single component on the CS component library, so you'll need to add a single `CyberstormProvider` at the root of the app which includes everything necessary. We'll need to do this for global CSS variables and such at least so it wouldn't be hard to bundle both.

TL;DR: Check if a global provider can be used, and if there are downsides to it. If it works equally well, then let's use that pattern as it's more efficient.

File: apps/cyberstorm-storybook/stories/components/Tooltip.stories.tsx (hunk 3 above)
Comment (Oksamies) [reply apps/cyberstorm-storybook/stories/components/Tooltip.stories.tsx:30]: Alright, I've added
`  max-width: 75vw;
  overflow-wrap: break-word;`
And I think that could be fine, it at least looks fine. I hope that is fine.

File: packages/cyberstorm/src/components/Tooltip.tsx (hunk 5 above)
Comment (Oksamies) [reply packages/cyberstorm/src/components/Tooltip.tsx:26]: Alright, so I've now added a `Providers.tsx` file which houses the `CyberstormProviders`. The `RadixTooltip.Provider` is defined in there.

Still I do have some performance hit worries with this bundling of all the providers together. To my understanding, all the components within the app will be force updated, if for example the tooltips child is updated. Probably not a problem, that cant be fixed later, but something to keep in mind.

TLDR; provider is now global and working 👍

**Review by MythicManiac — APPROVED** [review]
Looks good enough to me 👍

File: packages/cyberstorm/src/components/Tooltip.tsx [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+  sticky = "always",
+  avoidCollisions = true,
+  children,
+}) => {
+  return (
+    <RadixTooltip.Root>
+      <RadixTooltip.Trigger>{children}</RadixTooltip.Trigger>
+      <RadixTooltip.Portal>
+        <RadixTooltip.Content
+          className={`${styles.root}`}
```
Comment (MythicManiac) [inline packages/cyberstorm/src/components/Tooltip.tsx:30]: Why is string interpolation used here? You can just use the variable directly: `className={styles.root}`

**Review by MythicManiac — COMMENTED** [review]
Found one minor nitpick, but won't consider it a blocker

File: packages/cyberstorm/src/components/Tooltip.tsx (hunk 9 above)
Comment (VilppeRiskidev) [reply packages/cyberstorm/src/components/Tooltip.tsx]: Fixed this :D

File: packages/cyberstorm/src/components/Tooltip/Tooltip.tsx [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+// Cyberstorm Tooltip component
+
+export const Tooltip: React.FC<PropsWithChildren<TooltipProps>> = ({
+  content,
+  sideOffset = 5,
+  side = "right",
+  collisionPadding = 20,
+  sticky = "always",
+  avoidCollisions = true,
```
Comment (VilppeRiskidev) [inline packages/cyberstorm/src/components/Tooltip/Tooltip.tsx:22]: I'd remove these default values from here and move them to a Tooltip.defaultProps (same as in Button.tsx for example).

File: packages/cyberstorm-styles/src/styles/theme.css [hunk 11]
```diff
@@ -31,4 +31,6 @@
   --color-text--default: var(--color-lilac--lightest);
   --color-text--secondary: var(--color-lilac--lighter);
   --color-text--tertiary: var(--color-lilac--light);
+  --animation-duration-1: 400ms;
```
Comment (VilppeRiskidev) [inline packages/cyberstorm-styles/src/styles/theme.css:34]: --animation-duration-default or --animation-duration-medium would be better imo (then in the future we could add slower and faster verisions)

File: packages/cyberstorm/src/components/Tooltip/Tooltip.module.css [hunk 12]
```diff
@@ -0,0 +1,87 @@
+.root {
+  max-width: 75vw;
+
+  padding: var(--space--8) var(--space--12);
+  border-radius: var(--border-radius--8);
+  color: var(--color-text--default);
+
+  font-weight: 700;
+  font-size: var(--font-size--14);
```
Comment (VilppeRiskidev) [inline packages/cyberstorm/src/components/Tooltip/Tooltip.module.css:9]: I'd also add `font: var(--font-body);` here, as we pass straight up strings in the story and they look yucky.

File: packages/cyberstorm/src/components/Tooltip/Tooltip.tsx [hunk 13]
```diff
@@ -0,0 +1,43 @@
+import React, { PropsWithChildren, ReactNode } from "react";
+import * as RadixTooltip from "@radix-ui/react-tooltip";
+import styles from "./Tooltip.module.css";
+
+export interface TooltipProps {
+  content?: ReactNode;
```
Comment (VilppeRiskidev) [inline packages/cyberstorm/src/components/Tooltip/Tooltip.tsx:6]: if (and only if) the content is always a string, change the prop's type to string. If the content can be something else (as I think it can) maybe you could add a story to showcase how that looks/works. (For example pass something like a Title, TextInput and a Button wrapped in a div).

File: apps/cyberstorm-storybook/stories/components/Tooltip.stories.tsx [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+  component: Tooltip,
+} as ComponentMeta<typeof Tooltip>;
+
+const Template: ComponentStory<typeof Tooltip> = (args) => (
+  <div
+    style={{
+      display: "flex",
+      flexDirection: "column",
+      gap: "0.25em",
+      backgroundColor: "#7e26b6",
```
Comment (VilppeRiskidev) [inline apps/cyberstorm-storybook/stories/components/Tooltip.stories.tsx:16]: Btw, you can use the variables here too. :P

**Review by VilppeRiskidev — COMMENTED** [review]
Commented some stuff, consider them and fix accordingly when if reasonable.

File: packages/cyberstorm-styles/src/styles/theme.css (hunk 11 above)
Comment (Cyanideankst) [reply packages/cyberstorm-styles/src/styles/theme.css:35]: Valid point. Fixed it to be *-default 👌

File: apps/cyberstorm-storybook/stories/components/Tooltip.stories.tsx (hunk 14 above)
Comment (Cyanideankst) [reply apps/cyberstorm-storybook/stories/components/Tooltip.stories.tsx:16]: I was harder than I thought it would be. Did change them there 👍

File: packages/cyberstorm/src/components/Tooltip/Tooltip.tsx (hunk 10 above)
Comment (Oksamies) [reply packages/cyberstorm/src/components/Tooltip/Tooltip.tsx:22]: I'd leave these here, since we probably want to have the same expected behavior of the Tooltip everywhere.

Comment (VilppeRiskidev) [reply packages/cyberstorm/src/components/Tooltip/Tooltip.tsx:22]: They'd still exist, just in a different place, the functionality would be the same, but this would be in line with all of the other components. Those values can be overridden as well as with the other style. 🤔 Is there some other reason I'm not seeing? All in all just a small little thing, should not become a blocker. @Cyanideankst do whatever feels right 🤷

Comment (VilppeRiskidev) [reply packages/cyberstorm/src/components/Tooltip/Tooltip.tsx:22]: Got an answer through discord. Does not require any modifications.
